### PR TITLE
GDScript: Fix typing of lambda functions

### PR DESCRIPTION
**Pull request by vonagam**
While making PR about value returns in void functions I tested how it was working with lambda functions and was surprised to find that lambda was not complaining about anything, turns out return type was never set...

```gdscript
func test():
  var lambda := func() -> int:
    return 'NO' # should complain now
  print(lambda.call())
```

Changed `reduce_lambda` to use `resolve_function_signature` and `resolve_function_body` just like others (member functions, getters, setters). Besides setting return type there are also checks for shadowing arguments and unused ones. Also maybe cyclic thing will be relevant.

Added `p_is_lambda` to `resolve_function_signature` because parent definitions and constructor logic does not applies to them.

Removed case for `GDScriptParser::Node::FUNCTION` in `resolve_node` - it was unneeded.

If there is any specific snippets that should be added to tests let me know.

### CODE REVIEW

**Review by adamscott — APPROVED** [review]
Looks good to me. Congrats for fixing lambdas.
Though, I think more tests for typed lambdas should be done, as this is the first one using them.

**Comment by vonagam** [comment]
Wow, you are right, no typed lambdas in prior tests.

Ok, added more tests specifically about what this PR is about - now wrong return, no return, shadowing arg and unused arg do produce errors/warnings. And added a basic test with working typed lambdas.

**Review by anvilfolk — APPROVED** [review]
Didn't do a deep dive and I'm not particularly well versed in lambdas, but this all looks good to me and the tests are a good indication it's all working as expected :)

File: modules/gdscript/tests/scripts/analyzer/errors/lambda_shadowing_arg.out [hunk 1]
```diff
@@ -0,0 +1,6 @@
+GDTEST_OK
+>> WARNING
+>> Line: 4
+>> SHADOWED_VARIABLE
+>> The local function parameter "shadow" is shadowing an already-declared variable at line 1.
+shadow
```
Comment (vnen) [inline modules/gdscript/tests/scripts/analyzer/errors/lambda_shadowing_arg.out:6]: This and the other tests for warnings should be in the `warnings` folder. I did notice there's no such folder under `analyzer` and all of those are in `parser/warnings` (not sure why). So either you can move to `parser/warnings` and we sort them out later or add these to a new `analyzer/warnings` folder.

**Review by vnen — CHANGES_REQUESTED** [review]

**Comment by vnen** [comment]
Just a nitpick on the location of the tests, otherwise looks good.

File: modules/gdscript/tests/scripts/analyzer/errors/lambda_shadowing_arg.out (hunk 1 above)
Comment (vonagam) [reply modules/gdscript/tests/scripts/analyzer/errors/lambda_shadowing_arg.out]: Done.

**Review by vnen — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**